Build failure: otb (OTBMosaicHeaderTest1.cxx.o)

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

With the latest master

```
nix-build -A otb
```

### Can Hydra reproduce this build failure?

Please select the Hydra Status.

### Link to Hydra build job

_No response_

### Relevant log output

```console
In file included from /build/source/build/Modules/Core/Mosaic/test/OTBMosaicHeaderTest1.cxx:23:
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h: In constructor 'otb::Functor::RGB2LAB<TInput, TOutput>::RGB2LAB()':
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:61:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
   61 |     D1[0][0] = 1.0 / vcl_sqrt(3.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:61:22: note: (if you use '', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:62:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
   62 |     D1[1][1] = 1.0 / vcl_sqrt(6.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:63:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
   63 |     D1[2][2] = 1.0 / vcl_sqrt(2.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h: In member function 'TOutput otb::Functor::RGB2LAB<TInput, TOutput>::operator()(const TInput&) const':
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:108:26: error: there are no arguments to 'vcl_log' that depend on a template parameter, so a declaration of 'vcl_log' must be available []
  108 |     const double log10 = vcl_log(10);
      |                          ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:109:26: error: there are no arguments to 'vcl_log' that depend on a template parameter, so a declaration of 'vcl_log' must be available []
  109 |     lms[0][0]          = vcl_log(lms[0][0]) / log10;
      |                          ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:110:26: error: there are no arguments to 'vcl_log' that depend on a template parameter, so a declaration of 'vcl_log' must be available []
  110 |     lms[1][0]          = vcl_log(lms[1][0]) / log10;
      |                          ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:111:26: error: there are no arguments to 'vcl_log' that depend on a template parameter, so a declaration of 'vcl_log' must be available []
  111 |     lms[2][0]          = vcl_log(lms[2][0]) / log10;
      |                          ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h: In constructor 'otb::Functor::LAB2RGB<TInput, TOutput>::LAB2RGB()':
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:162:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
  162 |     D1[0][0] = 1.0 / vcl_sqrt(3.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:163:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
  163 |     D1[1][1] = 1.0 / vcl_sqrt(6.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:164:22: error: there are no arguments to 'vcl_sqrt' that depend on a template parameter, so a declaration of 'vcl_sqrt' must be available []
  164 |     D1[2][2] = 1.0 / vcl_sqrt(2.0);
      |                      ^~~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h: In member function 'TOutput otb::Functor::LAB2RGB<TInput, TOutput>::operator()(const TInput&) const':
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:207:17: error: there are no arguments to 'vcl_pow' that depend on a template parameter, so a declaration of 'vcl_pow' must be available []
  207 |     lms[0][0] = vcl_pow(10.0, lms[0][0]);
      |                 ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:208:17: error: there are no arguments to 'vcl_pow' that depend on a template parameter, so a declaration of 'vcl_pow' must be available []
  208 |     lms[1][0] = vcl_pow(10.0, lms[1][0]);
      |                 ^~~~~~~
/build/source/Modules/Core/Mosaic/include/otbMosaicFunctors.h:209:17: error: there are no arguments to 'vcl_pow' that depend on a template parameter, so a declaration of 'vcl_pow' must be available []
  209 |     lms[2][0] = vcl_pow(10.0, lms[2][0]);
      |                 ^~~~~~~
[ 34%] Built target OTBCloudDetectionHeaderTest1
[ 34%] Built target OTBTransformHeaderTest1
[ 34%] Built target OTBIndicesHeaderTest1
[ 34%] Built target OTBImageNoiseHeaderTest1
[ 34%] Linking CXX executable ../../../bin/OTBLabellingHeaderTest1
[ 34%] Linking CXX executable ../../../../bin/OTBContrastHeaderTest1
[ 34%] Linking CXX executable ../../../../bin/OTBMathParserHeaderTest1
make[2]: *** [Modules/Core/Mosaic/CMakeFiles/OTBMosaicHeaderTest1.dir/build.make:83: Modules/Core/Mosaic/CMakeFiles/OTBMosaicHeaderTest1.dir/test/OTBMosaicHeaderTest1.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:11804: Modules/Core/Mosaic/CMakeFiles/OTBMosaicHeaderTest1.dir/all] Error 2
```

### Additional context

_No response_

### System metadata

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.88, NixOS, 24.11 (Vicuna), 24.11.20250502.bf3287d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.14`
```

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/ot/otb/package.nix
+++ b/pkgs/by-name/ot/otb/package.nix
@@ -234,9 +234,17 @@ stdenv.mkDerivation (finalAttrs: {
     ./1-otb-swig-include-itk.diff
   ];
 
-  postPatch = (
-    "substituteInPlace Modules/Core/Wrappers/SWIG/src/python/CMakeLists.txt --replace-fail '''$''{ITK_INCLUDE_DIRS}' ${otb-itk}/include/ITK-${itkMajorMinorVersion}"
-  );
+  postPatch =
+    ''
+      substituteInPlace Modules/Core/Wrappers/SWIG/src/python/CMakeLists.txt \
+        --replace-fail ''\'''${ITK_INCLUDE_DIRS}' "${otb-itk}/include/ITK-${itkMajorMinorVersion}"
+    ''
+    # Add the header file "vcl_legacy_aliases.h", which defines the legacy vcl_* functions.
+    # This patch fixes the unreproducible build of OTB.
+    # See https://gitlab.orfeo-toolbox.org/orfeotoolbox/otb/-/issues/2484.
+    + ''
+      sed -i '/#include "vcl_compiler.h"/a #include "vcl_legacy_aliases.h"' Modules/Core/Mosaic/include/otbMosaicFunctors.h
+    '';
 
   nativeBuildInputs = [
     cmake

